pkg/parser/parser.go

```diff
@@ -196,9 +196,42 @@ func (p *Parser) parseStatement(scope *scope) Node {
 	return nil
 }
 
-func (p *Parser) parseAssignStatement(scope *scope) Node {
+func (p *Parser) parseAssignmentStatement(scope *scope) Node {
+	target := p.parseAssignable(scope)
+	tok := p.cur
+	if target == nil {
+		p.advancePastNL()
+		return nil
+	}
+	p.assertToken(lexer.ASSIGN)
+	p.advance()
+	value := p.parseTopLevelExpression(scope)
+	if value == nil {
+		p.advancePastNL()
+		return nil
+	}
+	if !target.Type().Accepts(value.Type()) {
+		msg := "'" + target.String() + "' accepts values of type " + target.Type().Format() + ", found " + value.Type().Format()
+		p.appendErrorForToken(msg, tok)
+	}
+	p.assertEOL()
 	p.advancePastNL()
-	return nil
+	return &Assignment{Token: tok, Target: target, Value: value}
+}
+
+func (p *Parser) parseAssignable(scope *scope) Node {
+	name := p.cur.Literal
+	p.advance()
+	v, ok := scope.get(name)
+	if !ok {
+		if _, ok := p.funcs[name]; ok {
+			p.appendError("function call must be parenthesized: (" + name + " ...)")
```

Comment:
This error doesn't seem right. Isn't this the case of assigning to a function name and not a variable? Shouldn't this error say something like "cannot assign to <name> as it is a function not a variable"?

I think this is because an assignable (lvalue) is also valid as a term (rvalue), but in each context the error message should be different. On the left-hand side, you cannot assign to a function. On the right-hand side, a function name must be in parens.

You may want to add a test case for this too - assigning to a function.

Or should this just be an "unknown variable" error? Did you decide that a variable can shadow a function? I can't remember.

Reply:
yes, thanks for picking this up.
variables cannot shadow functions. unknown variable error seems perfect! I will add the test.